luaui/Widgets/cmd_blueprint.lua

```diff
@@ -272,6 +274,70 @@ local function setSelectedBlueprintIndex(index)
 	end
 end
 
+local function isValidBlueprint(blueprint)
+	if blueprint == nil then
+		return false
+	end
+
+	if blueprint.hasInvalidUnits then
+		return false
+	end
+
+	local buildable, unbuildable = WG["api_blueprint"].getBuildableUnits(blueprint)
+
+	if buildable == 0 then
+		return false
+	end
+
+	return true
+end
+
+local function getNextFilteredBlueprintIndex(startIndex, checkStart)
+	local newIndex = startIndex or selectedBlueprintIndex or 1
+	local foundValid = isValidBlueprint(blueprints[newIndex])
+
+	if foundValid and checkStart then
+		return newIndex
+	end
+
+	for _ = 1, #blueprints do
```

Comment:
wouldn't this for loop back to the initial blueprint? is this intended?

Reply:
Another refactoring artifact :) I think with `foundValid`, it looks like we can skip one loop iteration. The original intent was to fall back to returning the original index, but now we return nil instead.

Reply:
Actually looking at it right now, and maybe in some situations you do want to loop everything so maybe better not to change this.

Reply:
Like, when you change selection, you always go to lastExplicitlySelectedBlueprintIndex-1, then do getNext. In that case if lastExplicitlySelectedBlueprintIndex-1 is actually the only valid one.

I did test it and that's the case so I can answer: yes looping the whole list is intended (and needed).

Reply:
In any case since both of us had a bit of confusion regarding its need, a comment would be in order to know it's intended behaviour of the function. "return next valid blueprint possibly looping back to the original startIndex"